pex/cli/commands/lock.py

```diff
@@ -119,6 +131,237 @@ def __call__(self, parser, namespace, value, option_str=None):
+        if command:
+            argv0 = command[0] if os.path.isabs(command[0]) else venv.bin_path(command[0])
+            command = (argv0,) + command[1:]
+        return SyncTarget(venv=venv, command=command)
+
+    @classmethod
+    def resolve_venv(
+        cls,
+        argv0,  # type str
+        *additional_args  # type: str
+    ):
+        # type: (...) -> Union[SyncTarget, Error]
+
+        argv0_path = None  # type: Optional[str]
+        if is_exe(argv0):
+            argv0_path = argv0
+        else:
+            path = os.environ.get("PATH")
+            if path:
+                for entry in path.split(os.pathsep):
+                    exe_path = os.path.abspath(os.path.join(entry, argv0))
+                    if is_exe(exe_path):
+                        argv0_path = exe_path
+                        break
+
+        if argv0_path:
+            venv = Virtualenv.enclosing(python=argv0_path)
+            if not venv:
+                try:
+                    venv = Virtualenv(os.path.dirname(os.path.dirname(argv0_path)))
+                except InvalidVirtualenvError as e:
+                    return Error(
+                        "Could not find a valid venv enclosing {argv0} to sync: {err}.\n"
+                        "Try explicitly specifying a venv to sync with `--venv`.".format(
+                            err=e, argv0=argv0
+                        )
+                    )
+            command = [argv0_path]
+            command.extend(additional_args)
+            return SyncTarget(venv=venv, command=tuple(command))
+
+        if os.path.isdir(argv0) and not additional_args:
+            try:
+                return SyncTarget(venv=Virtualenv(argv0))
+            except InvalidVirtualenvError as e:
+                return Error(
+                    "The directory at {path} is not a valid venv to sync: {err}.\n"
+                    "Try explicitly specifying a venv to sync with `--venv`.".format(
+                        err=e, path=argv0
+                    )
+                )
+
+        return Error(
+            "Could not determine a venv to sync after examining {argv0}.\n"
+            "Try explicitly specifying a venv to sync with `--venv`.".format(argv0=argv0)
+        )
+
+    venv = attr.ib()  # type: Virtualenv
+    command = attr.ib(default=None)  # type: Optional[Tuple[str, ...]]
+
+    def sync(
+        self,
+        distributions,  # type: Iterable[Distribution]
+        confirm=True,  # type: bool
+        retain_pip=False,  # type: bool
+    ):
+        # type: (...) -> Result
+
+        abs_venv_dir = os.path.realpath(self.venv.venv_dir)
+
+        existing_distributions_by_project_name = {
+            dist.metadata.project_name: dist for dist in self.venv.iter_distributions()
+        }  # type: Dict[ProjectName, Distribution]
+
+        to_remove = []  # type: List[Distribution]
+        to_install = []  # type: List[Distribution]
+        for distribution in distributions:
```

Comment:
Confirming my understanding of transitive dependencies of requirements (e.g. `explicit` requires `hidden`, and the command is something like `pex3 lock sync explicit --venv ...`): 

1. does `distributions` include all of them, not just the top level requirements?
2. I don't see explicit tests for these sort of transitive dependencies, either for updating lockfiles (including something transitioning from transitive-only to top-level requirement and vice versa) or for updating a venv. Have I missed some? If not, is that something worth testing or is it well covered elsewhere?

Reply:
1. Yes, distributions come from a full transitive lock resolve.
2. There are extensive tests of lock create and lock update in other files. This change just re-uses that 2 year old infra.

That said re 2, the transition you mention is probably interesting enough to test here.

Reply:
Ok, tested all of 2 in `test_sync_venv_transitive_to_direct_and_vice_versa`.